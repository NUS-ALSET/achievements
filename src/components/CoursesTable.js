import React, { Fragment } from "react";
import PropTypes from "prop-types";
import { isEmpty } from "react-redux-firebase";
import Button from "material-ui/Button";
import { Link } from "react-router-dom";
import Table, {
  TableBody,
  TableCell,
  TableHead,
  TableRow
} from "material-ui/Table";

import withStyles from "material-ui/styles/withStyles";
import { APP_SETTING } from "../achievementsApp/config";

const styles = theme => ({
  link: {
    textDecoration: "none"
  },
  button: {
    margin: theme.spacing.unit
  }
});

class CoursesTable extends React.PureComponent {
  static propTypes = {
    ownerId: PropTypes.string.isRequired,
    courses: PropTypes.any.isRequired,
    onDeleteCourseClick: PropTypes.func.isRequired,
    classes: PropTypes.object.isRequired
  };

  render() {
    const { onDeleteCourseClick, classes, courses, ownerId } = this.props;

    return (
      <Fragment>
        <Table>
          <TableHead>
            <TableRow>
              <TableCell>Course name</TableCell>
              <TableCell>Instructor</TableCell>
              <TableCell>Actions</TableCell>
            </TableRow>
          </TableHead>
          <TableBody>
            {isEmpty(courses) ? (
              <TableRow>
                <TableCell>Empty list</TableCell>
              </TableRow>
            ) : (
              Object.keys(courses).map(courseId => {
                return (
                  <TableRow key={courseId}>
                    <TableCell>{courses[courseId].name}</TableCell>
                    <TableCell>{courses[courseId].instructorName}</TableCell>
                    <TableCell>
                      <Link
                        to={`${APP_SETTING.basename}courses/${courseId}`}
                        className={classes.link}
                      >
                        <Button className={classes.button} raised>
                          View
                        </Button>
                      </Link>
                      {courses[courseId].owner === ownerId && (
                        <Button
                          className={classes.button}
                          onClick={() => onDeleteCourseClick(courseId)}
                          raised
                        >
                          Delete
                        </Button>
                      )}
                    </TableCell>
                  </TableRow>
                );
              })
            )}
          </TableBody>
        </Table>
      </Fragment>
    );
  }
}

export default withStyles(styles)(CoursesTable);
