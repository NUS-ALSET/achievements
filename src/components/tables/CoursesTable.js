import { Link } from "react-router-dom";
import { isEmpty } from "react-redux-firebase";
import IconButton from "@material-ui/core/IconButton";
import PropTypes from "prop-types";
import React, { Fragment } from "react";
import Table from "@material-ui/core/Table";
import TableBody from "@material-ui/core/TableBody";
import TableCell from "@material-ui/core/TableCell";
import TableHead from "@material-ui/core/TableHead";
import TableRow from "@material-ui/core/TableRow";

import SearchIcon from "@material-ui/icons/Search";
import EditIcon from "@material-ui/icons/Edit";
import DeleteIcon from "@material-ui/icons/Delete";
import LinearProgress from "@material-ui/core/LinearProgress";

import withStyles from "@material-ui/core/styles/withStyles";
import {
  courseRemoveDialogShow,
  courseShowNewDialog
} from "../../containers/Courses/actions";

const styles = theme => ({
  link: {
    textDecoration: "none"
  },
  button: {
    margin: theme.spacing.unit
  }
});

class CoursesTable extends React.PureComponent {
  static propTypes = {
    ownerId: PropTypes.string.isRequired,
    courses: PropTypes.any.isRequired,
    dispatch: PropTypes.func.isRequired,
    // onDeleteCourseClick: PropTypes.func.isRequired,
    classes: PropTypes.object.isRequired,
    fetchedCourses: PropTypes.bool
  };

  onDeleteCourseClick = course =>
    this.props.dispatch(courseRemoveDialogShow(course.id, course.name));
  onEditCourseClick = course =>
    this.props.dispatch(courseShowNewDialog(course));

  render() {
    const { classes, courses, ownerId } = this.props;

    if (!this.props.fetchedCourses) {
      if (
        this.props.fetchedCourses === null ||
        this.props.fetchedCourses === undefined
      ) {
        return <p>No courses available!</p>;
      }
      return (
        <Fragment>
          <br />
          <br />
          <LinearProgress />
        </Fragment>
      );
    }

    return (
      <Table>
        <TableHead>
          <TableRow>
            <TableCell>Course name</TableCell>
            <TableCell>Instructor</TableCell>
            <TableCell style={{ width: 200 }}>Actions</TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {isEmpty(courses) ? (
            <TableRow>
              <TableCell colSpan={5}>Empty list</TableCell>
            </TableRow>
          ) : (
            Object.keys(courses)
              .map(id => ({ ...courses[id], id }))
              .sort((a, b) => (a.createdAt > b.createdAt ? -1 : 1))
              .map(course => (
                <TableRow key={course.id}>
                  <TableCell>{course.name}</TableCell>
                  <TableCell>{course.instructorName}</TableCell>
                  <TableCell>
                    <Fragment>
                      <Link
                        className={classes.link}
                        to={`/courses/${course.id}`}
                      >
                        <IconButton aria-label="Open course">
                          <SearchIcon />
                        </IconButton>
                      </Link>
                      {course.owner === ownerId && (
                        <Fragment>
                          <IconButton
                            aria-label="Edit course"
                            onClick={() => this.onEditCourseClick(course)}
                          >
                            <EditIcon />
                          </IconButton>
                          <IconButton
                            aria-label="Delete course"
                            onClick={() => this.onDeleteCourseClick(course)}
                          >
                            <DeleteIcon />
                          </IconButton>
                        </Fragment>
                      )}
                    </Fragment>
                  </TableCell>
                </TableRow>
              ))
          )}
        </TableBody>
      </Table>
    );
  }
}

export default withStyles(styles)(CoursesTable);
