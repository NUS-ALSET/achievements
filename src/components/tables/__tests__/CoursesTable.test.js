/* eslint-disable no-magic-numbers */
import React from "react";
import { createShallow } from "@material-ui/core/test-utils";
import sinon from "sinon";
import renderer from "react-test-renderer";
import { BrowserRouter as Router } from "react-router-dom";
import TableCell from "@material-ui/core/TableCell";

import CoursesTable from "../CoursesTable";
import { TableBody, TableRow } from "@material-ui/core";

describe("<CoursesTable>", () => {
  let shallow;
  let mockDispatch;
  let courses;

  beforeEach(() => {
    mockDispatch = sinon.spy();
    shallow = createShallow({
      dive: true
    });
    courses = [
      {
        id: "test",
        name: "Test"
      },
      {
        id: "test2",
        name: "Test2"
      }
    ];
  });

  it("renders correctly when there are multiple courses", () => {
    const tree = renderer
      .create(
        <Router>
          <CoursesTable
            courses={courses}
            dispatch={mockDispatch}
            fetchedCourses={true}
            ownerId="-DjgKGJ12hjh_"
          />
        </Router>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should render empty list", () => {
    const wrapper = shallow(
      <CoursesTable
        courses={[]}
        dispatch={mockDispatch}
        fetchedCourses={true}
        ownerId="-DjgKGJ12hjh_"
      />
    );

    expect(
      wrapper
        .find(TableBody)
        .find(TableRow)
        .find(TableCell).length
    ).toEqual(1);
  });

  it("should render activities list", () => {
    const wrapper = shallow(
      <CoursesTable
        courses={courses}
        dispatch={mockDispatch}
        fetchedCourses={true}
        ownerId="-DjgKGJ12hjh_"
      />
    );

    // Expect the wrapper object to be defined
    expect(wrapper.find(TableBody)).toBeDefined();
    expect(wrapper.find(TableBody).find(TableRow)).toHaveLength(courses.length);
  });

  it("should render buttons for owner", () => {
    const courses = [
      {
        id: "test",
        name: "Test",
        owner: "-DjgKGJ12hjh_"
      }
    ];
    const wrapper = shallow(
      <CoursesTable
        courses={courses}
        dispatch={mockDispatch}
        fetchedCourses={true}
        ownerId="-DjgKGJ12hjh_"
      />
    );

    // Expect the wrapper object to be defined
    expect(wrapper.find(TableBody)).toBeDefined();
    expect(
      wrapper
        .find(TableBody)
        .find(TableRow)
        .find(TableCell)
    ).toHaveLength(3);
  });

  it("renders correctly when there are no courses", () => {
    const tree = renderer
      .create(
        <CoursesTable
          courses={[]}
          dispatch={mockDispatch}
          fetchedCourses={true}
          ownerId="-DjgKGJ12hjh_"
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders correctly when there are multiple courses", () => {
    const tree = renderer
      .create(
        <Router>
          <CoursesTable
            courses={courses}
            dispatch={mockDispatch}
            fetchedCourses={true}
            ownerId="-DjgKGJ12hjh_"
          />
        </Router>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
