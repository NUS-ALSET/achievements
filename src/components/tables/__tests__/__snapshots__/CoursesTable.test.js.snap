// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CoursesTable> renders correctly when there are multiple courses 1`] = `
<table
  className="MuiTable-root-3"
>
  <thead
    className="MuiTableHead-root-4"
  >
    <tr
      className="MuiTableRow-root-5 MuiTableRow-head-8"
    >
      <th
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-head-11"
        scope="col"
      >
        Course name
      </th>
      <th
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-head-11"
        scope="col"
      >
        Instructor
      </th>
      <th
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-head-11"
        scope="col"
        style={
          Object {
            "width": 200,
          }
        }
      >
        Actions
      </th>
    </tr>
  </thead>
  <tbody
    className="MuiTableBody-root-22"
  >
    <tr
      className="MuiTableRow-root-5"
    >
      <td
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-body-12"
      >
        Test2
      </td>
      <td
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-body-12"
      />
      <td
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-body-12"
      >
        <a
          className="CoursesTable-link-1"
          href="/courses/1"
          onClick={[Function]}
        >
          <button
            aria-label="Open course"
            className="MuiButtonBase-root-29 MuiIconButton-root-23"
            disabled={false}
            onBlur={[Function]}
            onContextMenu={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex="0"
            type="button"
          >
            <span
              className="MuiIconButton-label-28"
            >
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root-32"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
              </svg>
            </span>
            <span
              className="MuiTouchRipple-root-41"
            />
          </button>
        </a>
      </td>
    </tr>
    <tr
      className="MuiTableRow-root-5"
    >
      <td
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-body-12"
      >
        Test
      </td>
      <td
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-body-12"
      />
      <td
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-body-12"
      >
        <a
          className="CoursesTable-link-1"
          href="/courses/0"
          onClick={[Function]}
        >
          <button
            aria-label="Open course"
            className="MuiButtonBase-root-29 MuiIconButton-root-23"
            disabled={false}
            onBlur={[Function]}
            onContextMenu={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex="0"
            type="button"
          >
            <span
              className="MuiIconButton-label-28"
            >
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root-32"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
              </svg>
            </span>
            <span
              className="MuiTouchRipple-root-41"
            />
          </button>
        </a>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`<CoursesTable> renders correctly when there are multiple courses 2`] = `
<table
  className="MuiTable-root-3"
>
  <thead
    className="MuiTableHead-root-4"
  >
    <tr
      className="MuiTableRow-root-5 MuiTableRow-head-8"
    >
      <th
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-head-11"
        scope="col"
      >
        Course name
      </th>
      <th
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-head-11"
        scope="col"
      >
        Instructor
      </th>
      <th
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-head-11"
        scope="col"
        style={
          Object {
            "width": 200,
          }
        }
      >
        Actions
      </th>
    </tr>
  </thead>
  <tbody
    className="MuiTableBody-root-22"
  >
    <tr
      className="MuiTableRow-root-5"
    >
      <td
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-body-12"
      >
        Test2
      </td>
      <td
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-body-12"
      />
      <td
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-body-12"
      >
        <a
          className="CoursesTable-link-1"
          href="/courses/1"
          onClick={[Function]}
        >
          <button
            aria-label="Open course"
            className="MuiButtonBase-root-29 MuiIconButton-root-23"
            disabled={false}
            onBlur={[Function]}
            onContextMenu={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex="0"
            type="button"
          >
            <span
              className="MuiIconButton-label-28"
            >
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root-32"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
              </svg>
            </span>
            <span
              className="MuiTouchRipple-root-41"
            />
          </button>
        </a>
      </td>
    </tr>
    <tr
      className="MuiTableRow-root-5"
    >
      <td
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-body-12"
      >
        Test
      </td>
      <td
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-body-12"
      />
      <td
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-body-12"
      >
        <a
          className="CoursesTable-link-1"
          href="/courses/0"
          onClick={[Function]}
        >
          <button
            aria-label="Open course"
            className="MuiButtonBase-root-29 MuiIconButton-root-23"
            disabled={false}
            onBlur={[Function]}
            onContextMenu={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex="0"
            type="button"
          >
            <span
              className="MuiIconButton-label-28"
            >
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root-32"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
              </svg>
            </span>
            <span
              className="MuiTouchRipple-root-41"
            />
          </button>
        </a>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`<CoursesTable> renders correctly when there are no courses 1`] = `
<table
  className="MuiTable-root-3"
>
  <thead
    className="MuiTableHead-root-4"
  >
    <tr
      className="MuiTableRow-root-5 MuiTableRow-head-8"
    >
      <th
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-head-11"
        scope="col"
      >
        Course name
      </th>
      <th
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-head-11"
        scope="col"
      >
        Instructor
      </th>
      <th
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-head-11"
        scope="col"
        style={
          Object {
            "width": 200,
          }
        }
      >
        Actions
      </th>
    </tr>
  </thead>
  <tbody
    className="MuiTableBody-root-22"
  >
    <tr
      className="MuiTableRow-root-5"
    >
      <td
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-body-12"
        colSpan={5}
      >
        Empty list
      </td>
    </tr>
  </tbody>
</table>
`;
